game/graphics/pipelines/opengl.cpp

```diff
@@ -340,10 +342,20 @@ void render_game_frame(int game_width,
   {
     auto p = scoped_prof("wait-for-dma");
     std::unique_lock<std::mutex> lock(g_gfx_data->dma_mutex);
-    // note: there's a timeout here. If the engine is messed up and not sending us frames,
-    // we still want to run the glfw loop.
-    got_chain = g_gfx_data->dma_cv.wait_for(lock, std::chrono::milliseconds(50),
-                                            [=] { return g_gfx_data->has_data_to_render; });
+    // note: there's a timeout here, so imgui can still be responsive even if we never render
+    // anything. the timeout is extended to the equivalent of 0.5 frames per second (2 seconds per
+    // frame) once we get a game frame. this is the avoid obnoxious flickering if the game slows
+    // down too much but is still working. if that long timeout occurs, we fall back to the fast one
+    // since the game might have froze (or is unusably slow).
+    got_chain = g_gfx_data->dma_cv.wait_for(
+        lock,
+        g_gfx_data->got_data_once ? std::chrono::seconds(2) : std::chrono::microseconds(16667),
+        [=] { return g_gfx_data->has_data_to_render || g_gfx_data->force_vsync; });
+    if (g_gfx_data->force_vsync) {
+      got_chain = g_gfx_data->has_data_to_render;
+    } else {
+      g_gfx_data->got_data_once = got_chain;
+    }
```

Comment:
do you need to set `force_vsync = false` to prevent this from triggering again and again?

Reply:
accidentally removed that line